### ISSUE TITLE
[async] thumbnail spins infinitely if you add (1, x, y) layer to a viewer with a (z, y, x) layer 

### ISSUE THREAD

**Issue by psobolewskiPhD** [napari/napari#7070]
### 🐛 Bug Report

First noted here: https://github.com/ome/ome-zarr-py/pull/389
But ome-zarr isn't required.
If you have experimental async enabled and a 3D layer open and add a layer with a singleton z dim (1, y, x) then the thumbnail will spin forever when any slice but the first is shown in 2D mode (3D is fine). The 2nd layer can be image or labels, the behavior is the same.
This suggests that data is still being loaded.

Edit: In fact, it occurs whenever the slider is on a slice that is outside the range of a layer. So if layer1 is (20, 20, 20) and layer2 is (15, 20, 20) they will load ok, but moving the slider to >15 will result in the loading indicator.

If async is off (note using `NAPARI_ASYNC=0` doesn't work, see https://github.com/napari/napari/issues/6637 ), then the thumbnail is black when any slice other than 0 is shown.

Not sure if this is just a visual glitch with the animation or really a loading process issue.

### 💡 Steps to Reproduce

ensure async is on, then run:
```
import numpy as np
import napari

layer_one = np.random.random([20, 20, 20])
layer_two = np.random.random([1, 20, 20])
viewer = napari.Viewer()
viewer.add_image(layer_one)
viewer.add_image(layer_two)

napari.run()
```

### 💡 Expected Behavior

The thumbnail loading symbol should represent loading, so if the data is loaded then it should not be animating.

### 🌎 Environment

napari: 0.5.0a2.dev777+g04e7e3f2e.d20240704
Platform: macOS-14.5-arm64-arm-64bit
System: MacOS 14.5
Python: 3.12.4 | packaged by conda-forge | (main, Jun 17 2024, 10:13:44) [Clang 16.0.6 ]
Qt: 6.7.1
PyQt6: 
NumPy: 1.26.4
SciPy: 1.14.0
Dask: 2024.6.2
VisPy: 0.14.3
magicgui: 0.8.3
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.7
npe2: 0.7.6

OpenGL:
  - GL version:  2.1 Metal - 88.1
  - MAX_TEXTURE_SIZE: 16384
  - GL_MAX_3D_TEXTURE_SIZE: 2048

Screens:
  - screen 1: resolution 1680x1050, scale 2.0

Optional:
  - numba: 0.60.0
  - triangle not installed

Settings path:
  - /Users/piotrsobolewski/Library/Application Support/napari/napari-dev_a1eb8b76ba95fa16ad06e26097b46b8455dfbf0b/settings.yaml
Plugins:
  - napari: 0.5.0a2.dev777+g04e7e3f2e.d20240704 (81 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-ome-zarr: 0.5.3.dev40+gfd744fc (2 contributions)
  - napari-svg: 0.2.0 (2 contributions)

**Comment by psobolewskiPhD** [comment]
So:
```
viewer.layers[-1].loaded
```
returns `False` for the problem case. And that is what is used to trigger the animation in the thumbnail if I am reading https://github.com/napari/napari/pull/5342 correctly.
So it seems that the issue is related to setting `loaded` on the layer properly for the case where layers have different number of slices.

**Comment by psobolewskiPhD** [comment]
Looks like this code is working correctly:
https://github.com/napari/napari/blob/bf157e37685228609956168d8a71166ad40e874a/napari/layers/image/_slice.py#L327-L342
and
https://github.com/napari/napari/blob/bf157e37685228609956168d8a71166ad40e874a/napari/layers/image/_slice.py#L192-L196
But the latter doesn't change the `loaded` state?

**Comment by jni** [comment]
Oh wow, nice sleuthing @psobolewskiPhD! So I guess we just need to set loaded = True when out of bounds and we'll be golden???

**Comment by jni** [comment]
(In the future, we might want a dedicated indicator for out of bounds, but that requires quite a bit of UX/design research to get right, especially because points etc are out of bounds *most* of the time. 😂)

**Comment by psobolewskiPhD** [comment]
> Oh wow, nice sleuthing @psobolewskiPhD! So I guess we just need to set loaded = True when out of bounds and we'll be golden???

I can't actually figure out how Layer.loaded is changed though 😢

**Comment by jni** [comment]
...

```
$ git grep "\.loaded" | cat
napari/_qt/_tests/test_async_slicing.py:    assert layer.loaded
napari/_qt/_tests/test_async_slicing.py:        assert not layer.loaded
napari/_qt/_tests/test_async_slicing.py:    qtbot.waitUntil(lambda: layer.loaded)
napari/_qt/containers/qt_layer_model.py:        layer_loaded = layer.loaded
napari/_qt/containers/qt_layer_model.py:            layer_loaded = layer.loaded and not viewer_playing
napari/_qt/qt_viewer.py:                vispy_layer.events.loaded.connect(self._qt_poll.wake_up)
napari/_vispy/layers/labels.py:        if not self.layer.loaded:
napari/components/_layer_slicer.py:        This is needed to support ``Layer.loaded`` in async slicing.
napari/layers/base/base.py:            self.events.loaded()
napari/layers/labels/labels.py:        if self._updated_slice is None or not self.loaded:
napari/layers/labels/labels.py:        if not self.loaded:
napari/layers/utils/_tests/test_link_layers.py:    assert l1.loaded
napari/layers/utils/_tests/test_link_layers.py:    assert l2.loaded
napari/layers/utils/_tests/test_link_layers.py:    assert not l1.loaded
napari/layers/utils/_tests/test_link_layers.py:    assert l2.loaded
```

... me neither??? 😵‍💫 I'll fish around... probably some setattr somewhere...

**Comment by jni** [comment]
Ah, the plot thickens (I suspected looking in base would help):

```python
    def _set_loaded(self, loaded: bool) -> None:
        """Set the loaded state and notify a change with the loaded event."""
        if self._loaded != loaded:
            self._loaded = loaded
            self.events.loaded()
```

I'm not sure why it's not a standard property getter/setter but I'm sure @andy-sweet had a good reason... Which I hope he will share here in due course. 😂

**Comment by jni** [comment]
another clue! Maybe I should just work quietly and update when I have the full picture... 😂 

```
        This is private but accessed externally because it is related to slice
        state, which is intended to be moved off the layer in the future.
```

**Comment by jni** [comment]
Ok so the issue as far as I can tell is that `make_empty` *always* gives you a response with a new request id:

https://github.com/napari/napari/blob/374a6df7074176d1b5a2688360c55ebca000ed08/napari/layers/image/_slice.py#L117

That means that when you get to `_update_loaded_slice_id` here:

https://github.com/napari/napari/blob/374a6df7074176d1b5a2688360c55ebca000ed08/napari/layers/base/base.py#L698-L705

the slice id can *never* correspond to the `self._last_slice_id` (set in `set_unloaded_slice_id`, [just above](https://github.com/napari/napari/blob/374a6df7074176d1b5a2688360c55ebca000ed08/napari/layers/base/base.py#L689-L696) when making a new slice request), so the slice can never be set to loaded.

I *think* the fix is that `make_empty` should take a with_request_id argument optionally, which can be `None`by default (for when an empty slice is made during layer construction), in which case it can return a new one as done now, but which can also be set to `self.id` [inside the `if self._slice_out_of_bounds()` clause](https://github.com/napari/napari/blob/374a6df7074176d1b5a2688360c55ebca000ed08/napari/layers/image/_slice.py#L193-L196). That way the layer will know that the slice *is in fact empty* and it can stop its endless search for data. 🥹 

... But it would be good to get confirmation from @andy-sweet that this is a reasonable approach. 😅

**Comment by andy-sweet** [comment]
Thanks for investigating this. I should be able to take a look on Sunday.

**Comment by jni** [comment]
@andy-sweet too slow, it turns out! 😂 As punishment you have to review #7077 or even worse, fix #5316. 😜 

In all seriousness, I'm glad I took this one, gave me a little bit more practice with/understanding of the async machinery. 😊 Huge thanks and props to @psobolewskiPhD for the tiny reproducer and finding of the `make_empty` call, too, because I thought it was *definitely* gonna be an ome-ngff error / user error rather than a napari one!

**Comment by andy-sweet** [comment]
Thanks for all the sleuthing here. After a quick look, this seems like the right fix for now.

> another clue! Maybe I should just work quietly and update when I have the full picture... 😂
> 
> ```
>         This is private but accessed externally because it is related to slice
>         state, which is intended to be moved off the layer in the future.
> ```

Yeah, this is the main reason things are ugly here. Especially because we need to support having some slice state on layer initialization. We could make `Layer._slice` optional, but that may need checks in lots of places. I can imagine a future when slice state isn't on the layer, and might be on a dedicated store instead (e.g. related to a canvas) where I could just go look up the latest slice state related to a layer instead.

> Ah, the plot thickens (I suspected looking in base would help):
> 
> ```python
>     def _set_loaded(self, loaded: bool) -> None:
>         """Set the loaded state and notify a change with the loaded event."""
>         if self._loaded != loaded:
>             self._loaded = loaded
>             self.events.loaded()
> ```
> 
> I'm not sure why it's not a standard property getter/setter but I'm sure @andy-sweet had a good reason... Which I hope he will share here in due course. 😂

Like you discovered later, I wanted to avoid adding to the public API when this is stuff that will hopefully disappear soon. The `loaded` property getter already existed, but I didn't want to define a `loaded` property setter.
